azurelinuxagent/ga/exthandlers.py

```diff
@@ -1068,55 +1070,79 @@ def is_responsive(heartbeat_file):
-
-        ret = process.poll()
-        if ret is None:
-            raise ExtensionOperationError("Process {0} was not terminated: {1}\n{2}".format(process.pid, cmd, msg),
-                                          code=extension_error_code)
-        if ret != 0:
-            raise ExtensionOperationError("Non-zero exit code: {0}, {1}\n{2}".format(ret, cmd, msg),
-                                          code=extension_error_code)
+                try:
+                    # Some extensions erroneously begin cmd with a slash; don't interpret those
+                    # as root-relative. (Issue #1170)
+                    full_path = os.path.join(base_dir, cmd.lstrip(os.path.sep))
+
+                    def pre_exec_function():
+                        """
+                        Change process state before the actual target process is started. Effectively, this runs between
+                        the fork() and the exec() of sub-process creation.
+                        """
+                        os.setsid()
+                        CGroups.add_to_extension_cgroup(self.ext_handler.name)
+
+                    process = subprocess.Popen(full_path,
+                                               shell=True,
+                                               cwd=base_dir,
+                                               stdout=stdout,
+                                               stderr=stderr,
+                                               env=env,
+                                               preexec_fn=pre_exec_function)
+                except OSError as e:
+                    raise ExtensionOperationError("Failed to launch '{0}': {1}".format(full_path, e.strerror),
+                                                  code=extension_error_code)
 
-        duration = elapsed_milliseconds(begin_utc)
-        log_msg = "{0}\n{1}".format(cmd, "\n".join([line for line in msg.split('\n') if line != ""]))
-        self.logger.verbose(log_msg)
-        self.report_event(message=log_msg, duration=duration, log_event=False)
+                cg = CGroups.for_extension(self.ext_handler.name)
+                CGroupsTelemetry.track_extension(self.ext_handler.name, cg)
+                msg = ExtHandlerInstance._capture_process_output(process, stdout, stderr, cmd, timeout, extension_error_code)
+
+                duration = elapsed_milliseconds(begin_utc)
+                log_msg = "{0}\n{1}".format(cmd, "\n".join([line for line in msg.split('\n') if line != ""]))
+                self.logger.verbose(log_msg)
+                self.report_event(message=log_msg, duration=duration, log_event=False)
+
+                return msg
+
+    @staticmethod
+    def _capture_process_output(process, stdout_file, stderr_file, cmd, timeout, code=-1):
+        retry = timeout
+        while retry > 0 and process.poll() is None:
+            time.sleep(1)
+            retry -= 1
+
+        def read_output():
```

Comment:
As discussed, it would be a good idea to use try/catch for this method in the future.

Reply:
Thanks, @pgombar. Added error handling and also limited the amount of data we load into memory.